Estimate spoilage time in Eat menu and description (if skilled)

narc0tiq opened the pull request:
#### Summary
SUMMARY: Interface "Show estimated time until spoilage in Eat menu (if skilled)"

#### Purpose of change
This is #25262 rebased, squashed in places, [spindled, and mildly mutilated](https://idioms.thefreedictionary.com/fold%2C+spindle%2C+or+mutilate). Note this also includes #26318, which contains the ability to show numer of weeks remaining.

The original description from #25262 follows (with minimal edits):

> Being able to [estimate] how long the food you have will roughly last is very useful. Currently you only have the `(old)` and `(fresh)` tags to work with, but usually most of your food will have neither.

#### Describe the solution
Instead of the time food would last if newly created, the `E`at menu now shows the amount of time the food has left.

*It will only show the Freshness and Spoils In columns if you have Cooking 3 or Survival 4.* This is consistent with the detailed freshness estimation in the description of food.

* If the remaining time is less than or equal to 6 hours and the item is `(old)` it shows "soon!". This is to prevent having items with "6 hours" left from rotting. (maybe "<6 hours" instead of "soon!"/" 6 hours"  would be better?) 
* If the remaining time is less than or equal to 6 hours and the item is not `(old)`, it shows 6 hours.
* If the remaining time is less than or equal to 1 week, it shows the amount of days, rounded up.
* Otherwise it will round up to the next week.
* This will then be changed into seasons or years if appropriate.

#### Screenshots
Item descriptions for fresh food:
![fresh](https://i.imgur.com/9jXLVBc.png)
food past midlife:
![past midlife](https://i.imgur.com/Sp2XFU9.png)
and food that will be old soon:
![old soon](https://i.imgur.com/xNl7t92.png)
For an unskilled person, food that is neither fresh nor old looks like this:
![need more skill](https://user-images.githubusercontent.com/3521568/44909532-0d63c080-ad20-11e8-969e-f44d0960f983.PNG)

Eat menu for an unskilled person:
![unskilled eater](https://i.imgur.com/zONDwOZ.png)
And for a skilled person, the same tile:
![skilled eater](https://i.imgur.com/1AEUjjb.png)

nexusmrsep commented:
Shelf life of non perishables could say "infinite" to avoid confusion.

nexusmrsep commented:
Also one more thing we were discussing with @MT_Arnoldusen is slightly changing the perishable note in food description to emphasize that's a nominal shelf life. "This food is perishable and in room temperature has estimated nominal shelf life of (time)." Or something along those lines.

narc0tiq commented:
That makes perfect sense to me, I'll include that.

pikolinianita commented:
I have problem with flour in your example (in game too). I cannot see its freshness without enter item description. I can't even say if it is perishable or not. Is it possible to shorten "not edible in current form" message and, perhaps, display value in 9at least one of) new column?

BevapDin commented on src/game_inventory.cpp:
```diff
@@ -384,6 +387,32 @@ class comestible_inventory_preset : public inventory_selector_preset
+                    g->u.get_skill_level( skill_survival ) >= 4 ) {
+                    const islot_comestible item = get_edible_comestible( loc );
+                    if( item.spoils > 0 ) {
+                        std::string freshness = get_freshness( loc );
```
Is there a reason you store this in a local variable instead of returning it directly?

narc0tiq replied:
As I recall, it's for de-constifying, but for a very annoying reason -- we also return a plain `new std::string()` below, which is definitely non-const, and a function obviously can't have two return types...

There must be a better solution, but I'm not a C++ person and the google isn't helping. How do I get the `new std::string()` to be const?

jbytheway replied:
It should just work to `return get_freshness(loc)`.  I tried checking out your branch and that change compiles for me.

BevapDin replied:
> As I recall, it's for de-constifying, but for a very annoying reason -- we also return a plain new std::string() below, which is definitely non-const, and a function obviously can't have two return types...

> There must be a better solution, but I'm not a C++ person and the google isn't helping. How do I get the new std::string() to be const?

`return new std::string()` returns a pointer (to `std::string`). `return std::string()` returns a string value, neither is const.

An existing lambda is this:
```C++
append_cell( [ this ]( const item_location & loc ) {
    const time_duration spoils = get_edible_comestible( loc ).spoils;
    if( spoils > 0 ) {
        return to_string_clipped( spoils );
    }
    return std::string();
}, _( "SPOILS IN" ) );
```
It returns a new string value, or a string value from function call, basically e the same as in your code.

narc0tiq replied:
@BevapDin Yes, but `to_string_clipped` does not return a `const std::string`. Which suggests the correct solution is to make `get_freshness()` and friends to return a non-const, as well.

@jbytheway 
> that change compiles for me.

Sure, but `clang++` complains:

```
CCACHE_CPP2=1 ccache clang++  -DRELEASE -DGIT_VERSION -DLOCALIZE -ffast-math -Os  -Wall -Wextra  -g  -fsigned-char -Werror -std=c++11 -MMD -MP -m32 -I/usr/include/lua5.2 -DLUA -D_GNU_SOURCE -I/usr/include/ncursesw -c src/game_inventory.cpp -o obj/game_inventory.o
[...]
src/game_inventory.cpp:400:17: error: return type 'basic_string<[...]>' must match previous return type 'const basic_string<[...]>' when lambda expression has unspecified explicit return type
                return std::string();
                ^
1 error generated.
Makefile:787: recipe for target 'obj/game_inventory.o' failed
make: *** [obj/game_inventory.o] Error 1
```

Really, the solutions are simple: 
- have `get_freshness` return a non-const string
- de-const the `return get_freshness(...)`
- const the other `return std::string()`

I'm going to apply the first of these, please advise if you'd rather see something else.

narc0tiq replied:
@BevapDin Yes, but `to_string_clipped` does not return a `const std::string`. Which suggests the correct solution is to make `get_freshness()` and friends to return a non-const, as well.

@jbytheway 
> that change compiles for me.

Sure, but `clang++` complains:

```
CCACHE_CPP2=1 ccache clang++  -DRELEASE -DGIT_VERSION -DLOCALIZE -ffast-math -Os  -Wall -Wextra  -g  -fsigned-char -Werror -std=c++11 -MMD -MP -m32 -I/usr/include/lua5.2 -DLUA -D_GNU_SOURCE -I/usr/include/ncursesw -c src/game_inventory.cpp -o obj/game_inventory.o
[...]
src/game_inventory.cpp:400:17: error: return type 'basic_string<[...]>' must match previous return type 'const basic_string<[...]>' when lambda expression has unspecified explicit return type
                return std::string();
                ^
1 error generated.
Makefile:787: recipe for target 'obj/game_inventory.o' failed
make: *** [obj/game_inventory.o] Error 1
```

Really, the solutions are simple: 
- have `get_freshness` return a non-const string
- de-const the `return get_freshness(...)`
- const the other `return std::string()`

I'm going to apply the first of these, please advise if you'd rather see something else.

narc0tiq commented:
> I have problem with flour in your example (in game too). I cannot see its freshness without enter item description.

That's fair, I have a similar problem with frozen food -- yes, it won't spoil, but what would I get out of it (calories/quench) if I defrosted? I'll look into it, but I don't want to remove the feedback for why things are inedible.

codemime commented on src/game_inventory.cpp:
```diff
@@ -482,6 +511,52 @@ class comestible_inventory_preset : public inventory_selector_preset
+        const std::string get_time_left_rounded( const item_location &loc ) {
+            const item *item = loc.get_item();
+            const double relative_rot = item->is_food_container() ? item->contents.front().get_relative_rot() :
+                                        item->get_relative_rot();
```
This can (or rather should) be replaced with `get_comestible_item`.

narc0tiq replied:
Sounds good to me, thanks!

codemime commented on src/game_inventory.cpp:
```diff
@@ -482,6 +511,52 @@ class comestible_inventory_preset : public inventory_selector_preset
+            if( time_left > get_edible_comestible( loc ).spoils ) {
+                time_left = get_edible_comestible( loc ).spoils;
+            }
+            return to_string_clipped( time_left );
```
Can't it be replaced with `to_string_approx`? This could probably make all the conversions above unneeded:

```c++
if( time_left <= 6_hours && item->is_going_bad() ) {
    return _( "soon!" );
}

return to_string_approx( time_left );
```

Also, `std::ceil` is preferred to `ceilf`.

narc0tiq replied:
I'm not sure why the original code didn't use `to_string_approx`, but I just found out about it and I like it, so I'm going to use it.

![To String Approx](https://i.imgur.com/h35LHDw.png)

narc0tiq replied:
I'm not sure why the original code didn't use `to_string_approx`, but I just found out about it and I like it, so I'm going to use it.

![To String Approx](https://imgur.com/h35LHDw)

narc0tiq commented on src/game_inventory.cpp:
```diff
@@ -384,6 +387,32 @@ class comestible_inventory_preset : public inventory_selector_preset
+
+            append_cell( [this]( const item_location & loc ) {
+                if( g->u.get_skill_level( skill_cooking ) >= 3 ||
+                    g->u.get_skill_level( skill_survival ) >= 4 ) {
```
Agreed, this needs to be unified.

narc0tiq commented on src/game_inventory.cpp:
```diff
@@ -482,6 +511,52 @@ class comestible_inventory_preset : public inventory_selector_preset
+                return _( "near midlife" );
+            } else if( rot_progress >= 0.5 && rot_progress < 0.7 ) {
+                return _( "past midlife" );
+            } else if( rot_progress >= 0.7 && rot_progress < 0.9 ) {
```
Agreed, done.

narc0tiq commented:
Sorry about the email/notification spam, everyone, I didn't realize Github was going to remember all those POSTs and eventually add all the comments.

As I'd promised, I'm testing out an "Indefinite" marker for non-rottables' shelf life; it looks like this: ![Shelf Life: Indefinite](https://i.imgur.com/eRNsbzr.png)

And for a character that doesn't have the ability to estimate: ![Still indefinite](https://i.imgur.com/5uhlRuD.png)

I can't say it adds information for me, but it also doesn't look bad and should help discoverability of a feature (non-perishable foods), so I'll keep it.

nexusmrsep commented:
I like it more and more eche time I look at it. 
Few thoughts:
- I'm not sure it I like "Indefinite" green, but that might be a personal thing. Does it look better white to you?
- I would however de-capitalize it to "indefinite".
- I also like the Spoils in: "never" idea of yours. Or even a dash (or few) instead: "---" in both Freshness and Spoils in? 

Visual effect can't be properly judged without actually seeing them, so maybe try it and push it if you'll think it adds anything or just looks better then empty space. Thanks for considering.

narc0tiq commented:
> * I'm not sure it I like "Indefinite" green, but that might be a personal thing.

No, I agree, it looks weird. Maybe some other color would work better, but I think plain white/gray should be fine, plus _definitely_ the lowercase i and the "spoils in/never".

Let's take a look:
![white](https://i.imgur.com/rWCaZxq.png)
It doesn't look bad!
Here's another screenshot from a different page, next to some (sealed) tin cans:
![gray fresh](https://i.imgur.com/aODsRwd.png)

---

...now I'm wanting to sort by time-to-spoil, but I think that's nontrivial 😒

narc0tiq commented:
Oh, and here's the `---` option:
![dashes](https://i.imgur.com/WiG2uky.png)
![with some others](https://i.imgur.com/fp60jSH.png)

AMurkin commented:
I like indefinite and never variant.

narc0tiq commented:
> For distinguishing sealed foods, the white vs blue distinction on the times is not at all obvious to new players.

I agree, but I don't think we can fix that one in this PR (I have to draw the line somewhere, after all). This is actually a tricky problem and I'm not confident in a solution.

---

So, my own choice in marking permafoods is the dashes (with all three columns being dashes) -- it looks the least busy, especially when there are a *lot* of permafoods on screen. That said, it's the least obvious, whereas "indefinite + never" seems best in terms of being informative.

---

Here's some description examples, for fresh:
![fresh](https://i.imgur.com/9jXLVBc.png)
past midlife:
![past midlife](https://i.imgur.com/Sp2XFU9.png)
old soon:
![old soon](https://i.imgur.com/xNl7t92.png)

Not 100% impressed by the phrasing in some cases, I'll poke at it some more and push it in a few minutes.

narc0tiq commented:
Rebased, plus squashed the two commits that were adding the indefinite/never tags. I ended up going with "indefinite/---/never", being the most informative option.

I think this completes the WIP portion of this PR. I'm still open to changes in phrasing; I've added/changed several bits of text, and I'm sure someone has some suggestions to improve them.

nexusmrsep commented:
I'd personally go for "indefinite" for shelf life and "---" in other columns for the sole reason of it being informative. When using dashes only, the reason for why are they there is not obvious at a glance. You know that those columns are not applicable for a reason, but the reason is not exposed. You have to do the thought process to determine why some food items different. With "indefinite" you have a reason on a platter so the following dashes are obvious in an instant.

narc0tiq commented:
That sounds reasonable to me, I'll amend the last commit.

BevapDin commented on src/item.cpp:
```diff
@@ -760,6 +760,74 @@ std::string item::info( bool showtext, std::vector<iteminfo> &iteminfo, int batc
 
+// Generates a long-form description of the freshness of the given rottable food item.
+// NB: Doesn't check for non-rottable!
+const std::string get_freshness_description( const item *food_item )
```
Why return a const object?

`food_item` should be a reference.

narc0tiq replied:
> Why return a const object?

...habit? I'll fix it -- and the food_item. Give me a few minutes.

narc0tiq commented:
Not entirely on-topic, but I'd like to thank @BevapDin and @codemime for the detailed code reviews -- this project is the only time I work with C++ and I really appreciate the help in making sure the code is the best it possibly can.

narc0tiq commented:
@ZhilkinSerg: you're a true master of the merges nowadays -- I've seen your name on dozens of merged PRs. Thanks for the assist!

kevingranade commented:
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/more-details-in-consume-menu/22355/8
